### PR TITLE
Enforce choosing a delegate when click 'Delegate Support'

### PR DESCRIPTION
**Pull request by jinjagit**
Add feature: If user clicks 'Delegate Support' without selecting a specific delegate, prompt with a modal alert "Please first choose your delegate" that must be dismissed before continuing.

Relates to issue #97

### CODE REVIEW

**Comment by jinjagit** [comment]
Status:

I think I have the basic structure of this change here.

This works in Firefox and Chromium, but I need to test in other browsers (Chrome, Safari, Opera, etc.)

I also need to investigate an alternative to / workaround for `event.preventDefault();` as, apparently `event` in this context is now deprecated, but the code fails to work on Chromium without `event` as part of the command.

**Comment by jinjagit** [comment]
Actually, this works _until_ user successfully delegates, then undelegates, for example - then the prompt no longer appears as expected... needs more work.

**Comment by jinjagit** [comment]
Screenshot: Prompt if click 'Delegate Support' without selecting a delegate: 
prompt_delegate

**Comment by jinjagit** [comment]
This looks promising for my issues with losing the "You must choose a delegate" prompt functionality after ajax refresh : [reinitializing-javascript-after-ajax-refresh](This looks promising for my issues with losing the "You must choose a delegate" prompt functionality after ajax refresh : reinitializing-javascript-after-ajax-refresh)

File: app/views/decidim/proposals/proposal_votes/update_buttons_and_counters.js.erb [hunk 1]
```diff
@@ -17,6 +17,19 @@
     if($proposalVoteButton[0]) {
       morphdom($proposalVoteButton[0], '<%= j(render partial: "decidim/proposals/proposals/vote_button", locals: { proposal: proposal, from_proposals_list: @from_proposals_list }).strip.html_safe %>');
     }
+
+    $('#delegation-form').submit(function()
+    {
+      console.log($( "#select-delegate option:selected" ).text());
+
+      if ($( "#select-delegate option:selected" ).text() == "(choose delegate)") {
```
Comment (davefrey) [inline app/views/decidim/proposals/proposal_votes/update_buttons_and_counters.js.erb:25]: Don't forget to swap the hardcode with a `t(...)` translation, key is in en.yml

File: app/views/decidim/proposals/proposal_votes/update_buttons_and_counters.js.erb [hunk 2]
```diff
@@ -17,6 +17,19 @@
...
+        // 'event' in command below is, apparently, deprecated, but this does not work in Chromium
+        // without it. Needs more investigation.
+        event.preventDefault(); // This will prevent the form submit
+        alert('You must select a delegate!');
+        return false; // prevent the form submit (for IE)
```
Comment (davefrey) [inline app/views/decidim/proposals/proposal_votes/update_buttons_and_counters.js.erb:30]: I'm pretty bad on the front end, but I would think you don't need line 28, that it's `return false` that prevents form submission, just the conditional alert + return?

Comment (jinjagit) [reply app/views/decidim/proposals/proposal_votes/update_buttons_and_counters.js.erb:30]: I think you're right, the 'return false' is probably sufficient to prevent form submission for all browsers, I just haven't gotten around to checking this yet.  

Are you also suggesting we should not prevent form submission at all (by just using `return`, not `return false`)? This would not prevent the `ActiveRecord::RecordNotFound (Couldn't find Decidim::User with 'id'= [WHERE "decidim_users"."type" IN ('Decidim::User')])` when trying to delegate to no specified delegate.

File: app/assets/javascripts/decidim/liquidvoting/delegation.js.es6 [hunk 3]
```diff
@@ -0,0 +1,8 @@
+$(() => {
+  $('#delegation-form').submit(function() {
+    if ($( "#select-delegate option:selected" ).text() == "(choose delegate)") {
+      alert('You must choose a delegate!');
```
Comment (davefrey) [inline app/assets/javascripts/decidim/liquidvoting/delegation.js.es6:4]: I noticed these two string literals that are duplicated in the delegation partial, and needing translation which I'm guessing isn't done in assets, unsure though.

But a solution to both problems might be to put the strings into `t(...)` as `data:` attributes in the form in `_delegation.html.erb` (which gets translated), and then modify the two javascript files to reference the data in the DOM.

Comment (davefrey) [reply app/assets/javascripts/decidim/liquidvoting/delegation.js.es6:4]: Oh also, in the spirit of making decidim/delegation more fun and friendly, maybe "Please first choose a delegate" or some such :-)

**Comment by jinjagit** [comment]
The last commit removes code duplication as intended, but introduces a warning (browser console):
```
[Deprecation] Synchronous XMLHttpRequest on the main thread is deprecated because of its detrimental effects to the end user's experience. For more help, check https://xhr.spec.whatwg.org/.
```
My idea is to place the js code for the delegation modal in a file `delegation_modal.js.es6`, which is loaded in a partial `_delegation_modal.html.erb`, which is rendered at the end of the `_delegation.html.erb` partial. This means I can also use `append` in the `update_buttons_and_counters.js.erb` to re-render the `_delegation_modal.html.erb`, thus reloading the delegation_modal js code.  

This successfully avoids the previous approach of simply copying the entire contents of the `delegation_modal.js.es6` into `update_buttons_and_counters.js.erb`, which just feels wrong (and could become verbose as we add translation related functionality to the js code).

I have spent some time trying to research/fix the warning, and will continue to experiment. References include:
https://www.programmersought.com/article/12076948859/
https://thisinterestsme.com/synchronous-xmlhttprequest/
https://code2care.org/2015/synchronous-xmlhttprequest-main-thread-deprecated-detrimental-effect-end-users-experience

**Comment by jinjagit** [comment]
TLDR: This works, but uses a duplicate of code in the JS file `delegation_modal.js.es6`, first loaded in the the `_delegation.html.erb` partial, in the  `update_buttons_and_counters.js.erb` Ajax refresh code.  

I tried many approaches to try to reload the JS file contents in the Ajax refresh code, but all either did not work, or resulted in the warning about not being async mentioned in previous comments.  

There is a (relatively complex) modal template pattern used in Decidim core, which I haven't understood fully. However, I didn't find any usage of these modals that matched our use case (call modal if form submit contains certain form attribute value, &/or refresh the check for modal condition in Ajax refresh). Most (?all) are either opened via an explicit link (e.g. an info. modal) or due to a timeout (e.g. re-login prompt).

File: app/views/decidim/proposals/proposal_votes/update_buttons_and_counters.js.erb [hunk 4]
```diff
@@ -17,6 +17,15 @@
     if($proposalVoteButton[0]) {
       morphdom($proposalVoteButton[0], '<%= j(render partial: "decidim/proposals/proposals/vote_button", locals: { proposal: proposal, from_proposals_list: @from_proposals_list }).strip.html_safe %>');
     }
+
+    $('#delegation-form').submit(function() {
```
Comment (oliverbarnes) [inline app/views/decidim/proposals/proposal_votes/update_buttons_and_counters.js.erb:21]: I think I understand the need for this hack - re-rendering the delegation partial re-creates the delegation form in the DOM, and the `submit` event handler added by `delegation_modal.js.es6` is lost. 

And the `<%= javascript_include_tag "decidim/liquidvoting/delegation_modal" %>` at the bottom of the `_delegation.html.erb` partial doesn't get picked up during the re-render, possibly because the js gets escaped by `j()`.

But maybe this can be improved a bit:

- In the proposal listing page this will get executed once for each proposal in the list, because it's in the `@proposals.each` loop. We might be able to move this to the bottom of the script, ensuring it executes only once on js.erb render?

- Maybe we can replace the duplication with an include tag right here in the js.erb? This is more of a short in the dark, adapted from this hack for webpack: https://stackoverflow.com/questions/60945426/rails-6-how-to-include-javascript-pack-tag-in-js-erb-file

```javascript
(function() {
  $('body').append('<%= javascript_include_tag 'decidim/liquidvoting/delegation_modal' %>');
}());
```

**Review by oliverbarnes — CHANGES_REQUESTED** [review]
Looking good - I think I understand the need for the hack, and make a couple of suggestions to improve it and remove duplication

File: app/views/decidim/proposals/proposal_votes/update_buttons_and_counters.js.erb (hunk 4 above)
Comment (jinjagit) [reply app/views/decidim/proposals/proposal_votes/update_buttons_and_counters.js.erb:21]: >     * In the proposal listing page this will get executed once for each proposal in the list, because it's in the `@proposals.each` loop. We might be able to move this to the bottom of the script, ensuring it executes only once on js.erb render?


Good idea. Will do. 

>     * Maybe we can replace the duplication with an include tag right here in the js.erb? This is more of a short in the dark, adapted from this hack for webpack: https://stackoverflow.com/questions/60945426/rails-6-how-to-include-javascript-pack-tag-in-js-erb-file

I tried this approach (using `.append`) and it results in the async warning mentioned in comments here.

Comment (oliverbarnes) [reply app/views/decidim/proposals/proposal_votes/update_buttons_and_counters.js.erb]: > I tried this approach (using `.append`) and it results in the async warning mentioned in comments here.

If we're not actually making a synchronous ajax request, and it's simply a problem of having the script included in the response (https://stackoverflow.com/questions/24639335/javascript-console-log-causes-error-synchronous-xmlhttprequest-on-the-main-thr), then I think we can live with the warning. 

It's better than duplicating the script.

Would you also add some comments above the include please, explaining why we're having to do this?

Comment (jinjagit) [reply app/views/decidim/proposals/proposal_votes/update_buttons_and_counters.js.erb]: > If we're not actually making a synchronous ajax request, and it's simply a problem of having the script included in the response (https://stackoverflow.com/questions/24639335/javascript-console-log-causes-error-synchronous-xmlhttprequest-on-the-main-thr), then I think we can live with the warning.

Great, I'll do this :+1:

> Would you also add some comments above the include please, explaining why we're having to do this?

Will do :+1:

**Review by oliverbarnes — APPROVED** [review]

**Comment by davefrey** [comment]
Congrats for finding your way through that, even if there's a console warning it's a much better UX and system behavior
